Adds "Try it out" section for the simple GKE example. (#32)
GoogleCloudPlatform/opentelemetry-collector-builder-sample: deploy/gke/simple/README.md

## Try it out yourself (Optional)

Thus far, we've set up an OpenTelemetry collector configured to scrape itself and send metrics to the configured exporters. In this section we will introduce external source(s) of telemetry data to see how the collector operates on these sources. The sources will include data representing metrics, logs and traces.

You should be able to follow all steps mentioned in this README till [Verify the Deployment](#verify-the-deployment) successfully. This ensures that there are no permissions issues and you are able to succeesfully run a GKE cluster and connect to it.  

### Source of telemetry data

The example uses JSON file(s) containing telemetry data in OTLP format as source of telemetry data. These files can contain metrics, traces or logs - but each file should contain only a single type of telemetry data. Sample files have been provided in the folder - [otlp-data](./otlp-data/). The collector running in the cluster will then read these files and treat the data in them as if it were coming from a running application. 

### Updating the collector configuration to receive data from JSON file

> We need to update the collector configuration file - [otel-config](./otel-config.yaml) to add a reciever that is able to receive telemetry data from the added JSON file.

Reviewer: ```suggestion
We need to update the collector configuration [otel-config.yaml](./otel-config.yaml) file to add a reciever that is able to receive telemetry data from the added JSON file.
```

    start_at: "beginning"
    include:
      - "/mnt/testdata/metrics/*.json"
      - "/mnt/testdata/traces/*.json"
      - "/mnt/testdata/logs/*.json"
```
For more details about this particular receiver, check [otlpjsonfilereceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/otlpjsonfilereceiver).

***NOTE:** The path in `include` points to where the file would be mounted within the Kubernetes cluster environment and is therefore different from where it is present on your local machine. More information on this in [the next section](#making-the-json-file-available-within-the-cluster).*

Next, add this receiver in the traces pipeline, so your trace pipeline looks like - 

```
  traces:
    receivers: [otlp, otlpjsonfile]
    processors: [memory_limiter, batch, resourcedetection/gke]
    exporters: [googlecloud, logging]

  metrics:
    receivers: [otlp, prometheus/self, otlpjsonfile]
    processors: [memory_limiter, batch, resourcedetection/gke]
    exporters: [googlecloud, logging]
  
  logs:
    receivers: [otlp, otlpjsonfile]
    processors: [memory_limiter, batch, resourcedetection/gke]
    exporters: [googlecloud, logging]
```

> For exporting logs to Google Cloud Platform, we need to further configure the collector with the log name and the GCP project ID. The project ID is used by the cloud exporter to create GCP log entries. 

Reviewer: is project_id not inferred by the collector based on the GKE node it's running on?

Author: No, it was not since the logs were not being generated. I can double check again though by removing it and keeping the `default_log_name`.

Author: Just double checked, turns out that `default_log_name` was the required field in this case. Project ID is inferred.

            name: test-data-vol-logs      # This could be changed to anything
        ```  
    - Update the `spec.containers.volumeMounts` section to configure new volumeMount(s) for the added volumes.
        ```
        volumeMounts:
        - name: otel-collector-config-vol
          mountPath: /conf
        # Add the following volumeMount configurations
        - name: test-data-vol-metrics  # This name should match to the name of volume in spec.volumes
          mountPath: /mnt/testdata/metrics   # This is the location in the kubernetes environment where the file for configMap for metrics will be mounted
        - name: test-data-vol-traces  # This name should match to the name of volume in spec.volumes
          mountPath: /mnt/testdata/traces   # This is the location in the kubernetes environment where the file for configMap for traces will be mounted
        - name: test-data-vol-logs  # This name should match to the name of volume in spec.volumes
          mountPath: /mnt/testdata/logs   # This is the location in the kubernetes environment where the file for configMap for logs will be mounted
        ```

> 6. Deploy the cluster using the new deployment manifest - 

Reviewer: ```suggestion
6. Deploy the collector using the new deployment manifest - 
```

    ```
    kubectl apply -f manifest.yaml -n $OTEL_NAMESPACE
    ```
7. Verify if the telemetry data is now being emitted by using `kubectl logs` - 
    ```
    # Get the pod name(s)
    kubectl get pods -n $OTEL_NAMESPACE

    # Get the logs from a pod
    kubectl logs <pod_container_name> -n $OTEL_NAMESPACE
    ```

> *If the logs do not show the expected telemetry data, check if the file used to send json data - [testdata.json](./otlp-data/testdata.json) is found in the Kubernetes deployment. Follow steps in [troubleshooting guide](../troubleshooting.md#verify-updates-to-filesconfigmaps-in-kubernetes-cluster) to verify file presence and contents.*

Reviewer: I would cut this to just the last sentence
```
Follow steps in [troubleshooting guide](../troubleshooting.md#verify-updates-to-filesconfigmaps-in-kubernetes-cluster) to verify file presence and contents.
```

Author: Right, I will shorten it. I would prepend it with "In case there are any issues, " otherwise it might appear as a mandatory step.

> **NOTE: You will need to update the timestamps in [testdata-metrics](./otlp-data/testdata-metrics.json) to be within 24 hours of current time otherwise the metrics will not show up in Google Cloud console's Metrics Explorer. You might also need to update timestamps in other testdata - metrics and logs in case they become old enough to not be recognized by Google Cloud.**

Reviewer: is there any way we could automatically update these for the users with a Make command or something? Or give the user a time window to look at in their metrics explorer?

Author: I think we could use `sed` with `make` to automate this. Sounds like a good idea.

### Seeing the telemetry data

> At this point, we have several files acting as sources of telemetry data and a reciever configured withing the OpenTelemetry Collector that listens to these files whenever there are updates made to them. The collector also receives telemetry data from the files once when they are initially addedd (mounted).

Reviewer: ```suggestion
At this point, we have several files acting as sources of telemetry data and a receiver configured within the OpenTelemetry Collector that listens to these files whenever there are updates made to them. The collector also receives telemetry data from the files once when they are initially mounted.
```

1. Update the telemetry data in the desired data file - [testdata-metrics.json](./otlp-data/testdata-metrics.json), [testdata-traces.json](./otlp-data/testdata-traces.json). Make sure that JSON here is **not** pretty printed and remains minified. 
2. Update the corresponding ConfigMap. For instance, if you made changes to `testdata-traces.json` file, you need to update `otlp-test-data-traces` - 
    ```
    kubectl create configmap otlp-test-data-traces --from-file=./otlp-data/testdata-traces.json -n $OTEL_NAMESPACE --dry-run=client -o yaml | kubectl apply -f -
    ```

**NOTE**: It may take some time *(usually a few seconds)* for the config file to be updated.

To check if ConfigMap is updated, use following command with desired ConfigMap name - 
```
 kubectl describe configmaps otlp-test-data-traces -n $OTEL_NAMESPACE 
```

To check if the test file at mount path is updated, use following command with correct file path - 
```

>  kubectl exec -n $OTEL_NAMESPACE -it  <pod_container_name> -- cat /mnt/testdata/traces/testdata-traces.json 

Reviewer: this will only work if they're following the troubleshooting doc right? Maybe it should be moved there

Author: Yes, it does seem similar to the toubleshooting step but this is different from troubleshooting. 

This is in case the user wants to update their JSON data. Like it is possible that user reaches this step without needing to open troubleshooting, but I can remove this and link the troubleshooting section since content is same.

Reviewer: Yeah, the problem is that this command won't work unless they do the steps to update the Dockerfile and give them a shell in the container

Author: Oh yes, I see your point. I agree, better to move it there.
